src/IROperator.h

```diff
@@ -1562,20 +1545,35 @@ inline Expr operator<<(Expr x, int y) {
  * the type of the first argument. Both arguments must have integer
  * type. */
 // @{
+inline Expr operator<<(Expr x, Expr y) {
+    Type t = x.type();
+    return Internal::Call::make(t, Internal::Call::shift_left, {std::move(x), std::move(y)}, Internal::Call::PureIntrinsic);
+}
 inline Expr operator>>(Expr x, Expr y) {
-    Expr unsigned_amount = lossless_cast(UInt(x.type().bits(), y.type().lanes()), y);
-    user_assert(unsigned_amount.defined())
-        << "In shift right expression:\n"
-        << "   (" << x << ") >> (" << y << ")\n"
-        << "   with types " << x.type() << " >> " << y.type() << "\n"
-        << "the RHS must be unsigned and losslessly castable to the same size as the LHS.\n";
     Type t = x.type();
-    return Internal::Call::make(t, Internal::Call::shift_right, {std::move(x), std::move(unsigned_amount)}, Internal::Call::PureIntrinsic);
+    return Internal::Call::make(t, Internal::Call::shift_right, {std::move(x), std::move(y)}, Internal::Call::PureIntrinsic);
+}
+inline Expr operator<<(Expr x, int y) {
+    const bool flip_shift = y < 0;
```

Comment:
I think that because we should handle shifts of constant negative values in the simplifier, that we shouldn't *also* handle it here, which should make things simpler (only having this logic in one place).

Reply:
I tried to do that, but I either got cast errors going to unsigned t or segfaults later if I made t signed. Is this ok to handle it both places?

Reply:
I think we need to understand this better. I would be worried that if the simplifier produced a constant negative shift, that it would seg fault later too.

By removing this logic to handle this early, it would exercise the rest of the compiler with more tests (and avoid hiding future bugs), which needs to handle negative strides too.

Reply:
Updated now with the fix. Please take another look when you get a chance!